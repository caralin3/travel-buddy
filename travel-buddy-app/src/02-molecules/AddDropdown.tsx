import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { Dropdown, DropdownMenuItem } from '../01-atoms';
import {
  ADD_ACTIVITIES_ROUTE,
  ADD_CRUISES_ROUTE,
  ADD_FLIGHTS_ROUTE,
  ADD_HOTELS_ROUTE,
  ADD_PORTS_ROUTE,
  ADD_TRIPS_ROUTE,
} from '../router';

export interface AddDropdownProps {}

export const AddDropdown: React.FC<AddDropdownProps> = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const [dropdownOpen, setDropdownOpen] = React.useState(false);

  const showItem = (item: string) => !location.pathname.includes(item);

  const getItems = () => {
    const items: DropdownMenuItem[] = [];

    if (showItem('trips')) {
      items.push({
        label: 'Add Trip',
        onClick: () => navigate(ADD_TRIPS_ROUTE),
      });
    }
    if (showItem('flights')) {
      items.push({
        label: 'Add Flight',
        onClick: () => navigate(ADD_FLIGHTS_ROUTE),
      });
    }
    if (showItem('hotels')) {
      items.push({
        label: 'Add Hotel',
        onClick: () => navigate(ADD_HOTELS_ROUTE),
      });
    }
    if (showItem('cruises')) {
      items.push({
        label: 'Add Cruise',
        onClick: () => navigate(ADD_CRUISES_ROUTE),
      });
    }
    if (showItem('ports')) {
      items.push({
        label: 'Add Port',
        onClick: () => navigate(ADD_PORTS_ROUTE),
      });
    }
    if (showItem('activities')) {
      items.push({
        label: 'Add Activity',
        onClick: () => navigate(ADD_ACTIVITIES_ROUTE),
      });
    }
    return items;
  };

  return (
    <Dropdown isOpen={dropdownOpen} label="Add" menuItems={getItems()} toggle={() => setDropdownOpen(!dropdownOpen)} />
  );
};
