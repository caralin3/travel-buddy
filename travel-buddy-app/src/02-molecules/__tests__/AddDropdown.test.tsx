import { fireEvent, render, screen } from '@testing-library/react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { DropdownItemProps } from 'reactstrap';
import { AddDropdown } from '../AddDropdown';

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockedUseNavigate,
  useLocation: () => ({ pathname: 'trip' }),
}));

jest.mock('reactstrap', () => ({
  ...jest.requireActual('reactstrap'),
  DropdownItem: ({ onClick }: DropdownItemProps) => {
    const handleClick = (e: any) => {
      if (onClick) {
        onClick(e);
        mockedUseNavigate();
      }
    };
    return <button title="dropdown-item" onClick={handleClick}></button>;
  },
}));

describe('AddDropdown.tsx', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<AddDropdown />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should handle dropdown item click event', async () => {
    render(<AddDropdown />);
    const items = await screen.findAllByTitle('dropdown-item');
    for (let item of items) {
      fireEvent.click(item);
      expect(mockedUseNavigate).toHaveBeenCalled();
    }
  });
  it('should toggle dropdown', async () => {
    const { container } = render(<AddDropdown />);
    const toggleButton = container.querySelector('button.dropdown-toggle');
    if (toggleButton) {
      fireEvent.click(toggleButton);
      const expandedDropdown = container.querySelector('button.dropdown-toggle[aria-expanded="true"]');
      expect(expandedDropdown).toBeInTheDocument();
    }
  });
});
