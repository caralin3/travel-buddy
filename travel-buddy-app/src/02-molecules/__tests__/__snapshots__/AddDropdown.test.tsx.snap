// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddDropdown.tsx should render correctly 1`] = `
<Dropdown
  isOpen={false}
  label="Add"
  menuItems={
    Array [
      Object {
        "label": "Add Trip",
        "onClick": [Function],
      },
      Object {
        "label": "Add Flight",
        "onClick": [Function],
      },
      Object {
        "label": "Add Hotel",
        "onClick": [Function],
      },
      Object {
        "label": "Add Cruise",
        "onClick": [Function],
      },
      Object {
        "label": "Add Port",
        "onClick": [Function],
      },
      Object {
        "label": "Add Activity",
        "onClick": [Function],
      },
    ]
  }
  toggle={[Function]}
/>
`;
